add psc-package support

jimmyhuco opened the pull request:
BTW: purescript-prelude `^3.0.0` is enough.

jimmyhuco committed:
add psc-package support

BTW: purescript-prelude is only 3.0.0, not 3.1.0

kritzcreek commented on src/IdePurescript/Atom/Config.purs:
```diff
@@ -17,7 +17,7 @@ import Node.Platform (Platform(Win32))
 
 defaultSrcGlob :: Array String
-defaultSrcGlob = ["src/**/*.purs", "bower_components/**/*.purs"]
+defaultSrcGlob = ["src/**/*.purs", "bower_components/**/*.purs", ".psc-package/**/*.purs"]
```
One of the problems with this Glob is that it also picks up test files for dependencies, as well as multiple different package sets.

I'm using this Glob for my current project `".psc-package/psc-0.11.6/*/*/src/**/*.purs"`, but of course that only works if you're using the psc-0.11.6 package set.

I don't think this can just be a static string, since you need to analyze the current projects psc-package.json file.

nwolverson replied:
I agree, see comments on this issue: https://github.com/nwolverson/atom-ide-purescript/issues/170. You can of course use this glob in your own settings if it works for you.

Also and unrelated to this change, per @kRITZCREEK's point it looks like those bower paths are wrong too :|

kritzcreek replied:
@nwolverson Oh, yep. The globs we usually use:

`"src/**/*.purs" "bower_components/purescript-*/src/**/*.purs"`

kritzcreek replied:
Not that there are any packages that are not purescript left in bower_components usually, so that part can go, but the `src/` bit is relevant.

jimmyhuco replied:
Sorry, I have not tested this case carefully. I just try to use psc-package, but hyperclick can  not work. I add psc-package location, hyperclick works again. So I create this pull request.

Sorry. @nwolverson 
Thank you  @kRITZCREEK

jimmyhuco replied:
So the right way I think atom should check the project directory:

* if exist `bower_components`, then glob `"src/**/*.purs", "bower_components/purescript-*/src/**/*.purs"`
* elseif exist `.psc-package`, then check `psc-package.json` ,  to confirm which set  is  chosen now, then glob `"src/**/*.purs", ".psc-package/$SETNAME/*/*/src/**/*.purs"`

nwolverson replied:
I think I'd always prefer psc-package where present - I was thinking "if `psc-package.json` exists then add the output globs from `psc-package sources`".

jimmyhuco committed:
add feature: choosing psc-package or bower automatically

prefer psc-package

so if exist .psc-package, then check psc-package.json , to confirm
which set is chosen now, then glob "src/**/*.purs",
".psc-package/$SETNAME/*/*/src/**/*.purs"

elseif exist bower_components, then glob "src/**/*.purs", "bower_components/purescript-*/src/**/*.purs"

jimmyhuco commented:
@nwolverson I work out the idea above. You can check the logic is ok?

nwolverson commented:
Sorry for the delay in responding, but I'm not happy with this just now; I don't think this should be happening at startup time but instead as required (what happens when you add a package config file), and for that reason I won't go into more detailed comments.

I'd like to hold off on this for now, and think a little bit more about the best config setup for this scenario, with a view to moving to the shared language server etc. - as I don't think leaving this PR open blocks anyone (since you can add the glob in your local config if you want).

jimmyhuco commented:
Yes, it's not best plan. I can change glob in local config.
